substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+	///  the `SortedListProvider` is implemented by bags list due to score bucketing. Thus, we
+	///  leverage the [`SortedListProvider::in_position`] to verify if the target is in the
+	/// correct  position in the list (bag or otherwise), given its score.
+	pub fn do_try_state_voter_sorting() -> Result<(), sp_runtime::TryRuntimeError> {
+		for t in T::VoterList::iter() {
+			frame_support::ensure!(
+				T::VoterList::in_position(&t).expect("voter exists"),
+				"voter list is not sorted"
+			);
+		}
+
+		Ok(())
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	/// When a nominator's stake is updated, all the nominated targets must be updated
+	/// accordingly.
+	///
+	/// Note: it is assumed that `who`'s staking ledger state is updated *before* this method is
+	/// called.
+	fn on_stake_update(
+		who: &T::AccountId,
+		prev_stake: Option<Stake<BalanceOf<T>>>,
+		stake: Stake<BalanceOf<T>>,
+	) {
+		#[cfg(any(test, feature = "try-runtime))]"))]
+		debug_assert!(T::Staking::stake(who).unwrap() == stake, "input stake mismatches expected.");
+
+		match T::Staking::status(who) {
+			Ok(StakerStatus::Nominator(nominations)) => {
+				let voter_weight = Self::weight_of(stake.active);
+
+				let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+					"staker should exist in VoterList, as per the contract \
+                            with staking.",
+				);
+
+				let stake_imbalance = StakeImbalance::from(
+					prev_stake.map_or(Default::default(), |s| Self::to_vote_extended(s.active)),
+					voter_weight.into(),
+				);
+
+				// updates vote weight of nominated targets accordingly. Note: this will
+				// update the score of up to `T::MaxNominations` validators.
+				for target in nominations.into_iter() {
+					Self::update_target_score(&target, stake_imbalance);
+				}
+			},
+			Ok(StakerStatus::Validator) => {
+				let voter_weight = Self::weight_of(stake.active);
+				let stake_imbalance = StakeImbalance::from(
+					prev_stake.map_or(Default::default(), |s| Self::to_vote_extended(s.active)),
+					voter_weight.into(),
+				);
+
+				Self::update_target_score(who, stake_imbalance);
+
+				// validator is both a target and a voter.
+				let _ = T::VoterList::on_update(who, voter_weight).defensive_proof(
+					"the staker should exist in VoterList, as per the \
+                            contract with staking.",
+				);
+			},
+			Ok(StakerStatus::Idle) => (), // nothing to see here.
+			Err(_) => {
+				defensive!(
+					"staker should exist when calling `on_stake_update` and have a valid status"
+				);
+			},
+		}
+	}
+
+	/// A validator is also considered a voter with self-vote and should also be added to
+	/// [`Config::VoterList`].
+	//
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* calling this
+	/// method.
+	fn on_validator_add(who: &T::AccountId, self_stake: Option<Stake<BalanceOf<T>>>) {
```

Comment:
not clear to me at an interface level why `self_stake` can ever be an `Option`?

Reply:
It is probably over-engineering things, but the reasoning is that a validator may not have self-stake. That's not the case for our staking system, but rather for systems to use. Option can be dropped through and no self-stake can be  represented as `self_stake: 0` though.

Reply:
Yeah the latter seems better to me.